src/easylink/devtools/implementation_creator.py

```diff
@@ -36,22 +38,74 @@ class ImplementationCreator:
 
     def __init__(self, script_path: Path) -> None:
         self.script_path = script_path
+        self.recipe_path = script_path.with_suffix(".def")
+        self.container_path = script_path.with_suffix(".sif")
         self.implementation_name = script_path.stem
         self.requirements = self._extract_requirements(script_path)
         self.step_name = self._extract_step_name(script_path)
 
     def create_recipe(self) -> None:
         """Builds the singularity recipe and writes it to disk."""
 
-        recipe = PythonRecipe(self.script_path, self.requirements)
+        recipe = PythonRecipe(self.script_path, self.recipe_path, self.requirements)
         recipe.build()
         recipe.write()
         pass
 
     def build_container(self) -> None:
-        """Builds the container from the recipe."""
+        """Builds the container from the recipe.
+
+        Raises
+        ------
+        subprocess.CalledProcessError
+            If the container build fails.
+        FileNotFoundError
+            If singularity is not installed or not found in the system PATH.
+        """
         logger.info(f"Building container for '{self.implementation_name}'")
-        pass
+        if self.container_path.exists():
+            logger.warning(f"Container {self.container_path} already exists. Overwriting it.")
+
+        # Change to the recipe directory
+        os.chdir(self.container_path.parent)
+        try:
+            cmd = [
+                "singularity",
+                "build",
+                "--remote",
+                "--force",
+                str(self.container_path),
+                str(self.recipe_path),
+            ]
+            process = subprocess.Popen(
+                cmd,
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
+                text=True,
+            )
+            # stream output to console
+            if process.stdout:
+                for line in process.stdout:
+                    print(line, end="")
+            if process.stderr:
+                for line in process.stderr:
+                    print(line, end="")
+            process.wait()
+
+            if process.returncode == 0:
+                logger.info(f"Successfully built container '{self.container_path.name}'")
+            else:
+                logger.error(
+                    f"Failed to build container '{self.container_path.name}'. "
+                    f"Error: {process.returncode}"
+                )
+                raise subprocess.CalledProcessError(process.returncode, cmd)
+        except FileNotFoundError:
```

Comment:
Not super important, but I would be surprised if this were the exception raised in this situation